[leslieyip02] iP

leslieyip02 opened the pull request:
# Alice

Alice is a generic name that is commonly used when discussing cryptography. ([relevant xkcd](https://xkcd.com/1323/))

> Alice and Bob want to exchange a message or cryptographic key. Alice wants to send a private message to Bob, ...

## Features
- [x] Adding tasks
- [x] Removing tasks
- [x] Marking tasks as complete/incomplete
- [x] Listing tasks
- [x]  Saving tasks
- [ ] Interactive GUI

## Changes
- Rename chatbot
- Add `command`, `parser`, `storage`, `task` and `alice` packages

### Types of Tasks
1. To-Dos 👍
2. Deadlines 🎯
3. Events 📆

## Setup

- Download the `.jar` file [here](https://github.com/leslieyip02/ip/releases/tag/A-Jar)

OR

- Clone the repository and run `Alice::main`:

```
public class Alice {
    ...

    public static void main(String[] args) {
        Alice alice = new Alice();

        alice.greet();
        alice.listen();
        alice.bye();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

leslieyip02 committed:
Rename chatbot

leslieyip02 committed:
Add user input echo

leslieyip02 committed:
Add task tracking

leslieyip02 committed:
Add task completion tracking

leslieyip02 committed:
Add ToDo, Deadline and Event

ToDo, Deadline and Event are subclasses of Task.

ToDo have no time bounds.
Deadlines have a completion by field.
Events have from and to fields.

leslieyip02 committed:
Add basic test

leslieyip02 committed:
Add basic error handling

Handle invalid task string arguments.
Handle invalid task marking.

leslieyip02 committed:
Add task deletion

leslieyip02 committed:
Add delete test

leslieyip02 committed:
Add Commands enum

leslieyip02 committed:
Add save feature

leslieyip02 committed:
Change date string to LocalDateTime value

leslieyip02 committed:
Refactor UI into Ui class

leslieyip02 committed:
Refactor storage into Storage class

leslieyip02 committed:
Refactor input parser into Parser class

leslieyip02 committed:
Refactor task management into TaskList class

leslieyip02 committed:
Refactor commands into Command class

leslieyip02 committed:
Refactor into packages

leslieyip02 committed:
Update tests

leslieyip02 committed:
Merge branch 'add-gradle-support'

leslieyip02 committed:
Add ParserTest

leslieyip02 committed:
Add AddTaskTest

leslieyip02 committed:
Standardize imports

leslieyip02 committed:
Add DeleteTaskTest

leslieyip02 committed:
Configure JAR build

leslieyip02 committed:
Update for coding standard compliance

leslieyip02 committed:
Add JavaDoc

leslieyip02 committed:
Add FindTask command

leslieyip02 committed:
Merge branch 'branch-A-CodingStandard'

leslieyip02 committed:
Merge branch 'branch-Level-9'

leslieyip02 committed:
Refactor Parser class

somethingfishyfishy commented on src/main/java/alice/Alice.java:
```diff
@@ -0,0 +1,65 @@
+
+    private final Ui ui;
+    private final Storage storage;
+    private final TaskList taskManager;
```
surprising that the naming for this is taskManager

Does this mean that the TaskList class is doing more than managing a List<Task> Object?

somethingfishyfishy commented on src/main/java/alice/Alice.java:
```diff
@@ -0,0 +1,65 @@
+        ui.say(String.format("%s", line));
+    }
+
+    private void listen() {
```
Perhaps a better naming for this method would be run(), or loop()

Since you are not only listening for the command, but executing the command when it is not 'bye'

somethingfishyfishy commented on src/main/java/alice/ui/Ui.java:
```diff
@@ -0,0 +1,31 @@
+package alice.ui;
+
+/** Handles output to the user. */
+public class Ui {
```
I noticed that you always create a new Ui Object and store it as an attribute for other classes. I'm not saying it is wrong.

But you could make the below methods static methods so you don't require a Ui object to say things.

somethingfishyfishy commented on src/main/java/alice/command/Command.java:
```diff
@@ -0,0 +1,75 @@
+        if (input.startsWith(Action.TODO.name()) ||
+            input.startsWith(Action.DEADLINE.name()) ||
+            input.startsWith(Action.EVENT.name())) {
+            return new AddTask(ui, taskList);
```
I would think that AddTask (and the related) sounds more like a method name than a class name.

But there is nothing wrong with this currently.

somethingfishyfishy commented on src/main/java/alice/parser/DateParser.java:
```diff
@@ -0,0 +1,45 @@
+            try {
+                return LocalDateTime.parse(dateString, formatter);
+            } catch (DateTimeParseException exception) {
+                // try another format
```
perhaps print something, even if you're not doing anything with the exception object caught

somethingfishyfishy commented on src/main/java/alice/task/InvalidTaskException.java:
```diff
@@ -0,0 +1,11 @@
+package alice.task;
+
+public class InvalidTaskException extends Exception {
```
any javadoc comment?

leslieyip02 committed:
Configure checkstyle

leslieyip02 committed:
Fix style issues

leslieyip02 committed:
Migrate to JavaFX

leslieyip02 committed:
Use varargs

yugan01 commented on src/main/java/alice/command/DeleteTask.java:
```diff
@@ -0,0 +1,47 @@
+     * @param input the input line from the user
+     */
+    @Override
+    public String execute(String input) throws InvalidTaskException {
+        String[] tokens = input.split(" ", 2);
+        if (tokens.length != 2) {
+            throw new InvalidTaskException("Missing task number. Usage: delete <task number>");
+        }
+
+        int taskNumber;
+        try {
+            taskNumber = Integer.parseInt(tokens[1]);
+        } catch (NumberFormatException exception) {
+            throw new InvalidTaskException("Invalid task number. Usage: delete <task number>");
+        }
```
Should this be extracted out? As i observe the same code being repeated for DeleteTask/MarkTask and UnMarkTask

yugan01 commented on src/main/java/alice/parser/DateParser.java:
```diff
@@ -0,0 +1,45 @@
+/** Handles datetime and string conversions. */
+public class DateParser {
+    private static final DateTimeFormatter[] DATETIME_FORMATTERS = new DateTimeFormatter[]{
+        DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"),
+        DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm"),
+        DateTimeFormatter.ofPattern("dd-MM-yyyy HH:mm"),
+        DateTimeFormatter.ofPattern("dd-MM-yyyy HHmm"),
```
I like how you have a separate class to parse dates of varying formats.

yugan01 commented on src/main/java/alice/storage/TaskList.java:
```diff
@@ -0,0 +1,121 @@
+     * @param  task        the task to add
+     * @throws IOException if the tasks cannot be saved
+     */
+    public void addTask(Task task) throws IOException {
+        tasks.add(task);
+        saveTasks();
```
I understand that saveTasks() is called whenever a task is modified (added, deleted, marked). Is there any reason why saveTasks is not called at the end, when user says goodbye?

yugan01 reviewed (commented):
LGTM, your code was easy to follow and understand. I noticed a few situations where abstraction could be beneficial.

leslieyip02 committed:
Add assertions for tasks

leslieyip02 committed:
Add assertions for storage

leslieyip02 committed:
Add assertions for parsers

leslieyip02 committed:
Add assertions for commands

Add assertions for tasks, storage, parsers and commands. This checks
the validity of data within the program, and ensures that the bot
is working as intended.

leslieyip02 committed:
Improve code quality

Renamed variables for clarity. The variable names are now easier to
understand.

Removed storeTasks method in TaskList to adhere to SLAP.

leslieyip02 committed:
Refactor TaskList::findTasks to use streams

Use streams instead of for loops to simplify the code.

Streams were already used in FindTask::execute and ListTask::execute.

leslieyip02 committed:
Merge pull request #1 from leslieyip02/branch-A-Assertions

Add assertions

leslieyip02 committed:
Merge branch 'master' into branch-A-CodeQuality

leslieyip02 committed:
Merge branch 'master' into branch-A-Streams

leslieyip02 committed:
Merge pull request #2 from leslieyip02/branch-A-CodeQuality

Improve code quality

leslieyip02 committed:
Merge branch 'master' into branch-A-Streams

leslieyip02 committed:
Merge pull request #3 from leslieyip02/branch-A-Streams

Refactor TaskList::findTasks to use streams

leslieyip02 committed:
Add CI

leslieyip02 committed:
Add tag command

Tagging allows users to tag their tasks with text descriptions.

This commit also updates the toString method for tasks, as well as the
implementation for JSON serialization to accomodate the new field.

leslieyip02 committed:
Fix test failures

leslieyip02 committed:
Remove IO redirection tests

leslieyip02 committed:
Fix typo

leslieyip02 committed:
Update photos

leslieyip02 committed:
Update title of app

leslieyip02 committed:
Add representative screenshot

leslieyip02 committed:
Add user guide

leslieyip02 committed:
Fix jar build

leslieyip02 committed:
Fix code quality based on bot feedback

leslieyip02 committed:
Fix README img

leslieyip02 committed:
Fix Java Runtime version
